### PR TITLE
add support for soft dependencies

### PR DESCRIPTION
**Pull request by kdmcser**
to solve this problem discussed in modding channel of discordL
image
add soft dependency for a mod. The soft dependency mod can be not exist, but if exist, this mod will load after them(replace them).

### CODE REVIEW

File: lib/modding/CModHandler.cpp [hunk 1]
```diff
@@ -130,9 +137,11 @@ std::vector <TModID> CModHandler::validateAndSortDependencies(std::vector <TModI
 		if(!resolvedOnCurrentTreeLevel.empty())
 		{
 			resolvedModIDs.insert(resolvedOnCurrentTreeLevel.begin(), resolvedOnCurrentTreeLevel.end());
-					continue;
+			for(auto it = resolvedOnCurrentTreeLevel.begin(); it != resolvedOnCurrentTreeLevel.end(); it++)
+				notResolvedModIDs.erase(*it);
+			continue;
 		}
-		// If there're no valid mods on the current mods tree level, no more mod can be resolved, should be end.
```
Comment (kdmcser) [inline lib/modding/CModHandler.cpp:135]: fix the syntax error which I left three years ago. ^_^

File: lib/modding/CModHandler.cpp [hunk 2]
```diff
@@ -130,9 +137,11 @@ std::vector <TModID> CModHandler::validateAndSortDependencies(std::vector <TModI
...
+		// If there are no valid mods on the current mods tree level, no more mod can be resolved, should be ended.
```
Comment (kdmcser) [inline lib/modding/CModHandler.cpp:144]: fix the syntax error which I left three years ago. ^_^

File: lib/modding/CModHandler.cpp [hunk 3]
```diff
@@ -87,8 +87,9 @@ std::vector <TModID> CModHandler::validateAndSortDependencies(std::vector <TModI
 	std::vector <TModID> sortedValidMods; // Vector keeps order of elements (LIFO)
 	sortedValidMods.reserve(modsToResolve.size()); // push_back calls won't cause memory reallocation
 	std::set <TModID> resolvedModIDs; // Use a set for validation for performance reason, but set does not keep order of elements
+	std::set <TModID> notResolvedModIDs(modsToResolve.begin(), modsToResolve.end()); // Use a set for validation for performance reason
 
-	// Mod is resolved if it has not dependencies or all its dependencies are already resolved
+	// Mod is resolved if it has no dependencies or all its dependencies are already resolved
```
Comment (kdmcser) [inline lib/modding/CModHandler.cpp:92]: fix the syntax error which I left three years ago. ^_^

File: config/schemas/mod.json [hunk 4]
```diff
@@ -122,6 +122,11 @@
 			"description" : "List of mods that are required to run this one",
 			"items" : { "type" : "string" }
 		},
+		"softDepends" : {
+			"type" : "array",
+			"description" : "List of mods if exist, it should be loaded before this one",
```
Comment (krs0) [inline config/schemas/mod.json:127]: If I understood correctly, the soft dependency must be manualy enabled?

"List of mods where each mod, only if enabled, must be loaded ahead of this mod"

Comment (Olegmods) [reply config/schemas/mod.json:127]: > If I understood correctly, the soft dependency must be manualy enabled?
> 
> "List of mods where each mod, only if enabled, must be loaded ahead of this mod"

From what I understand, the soft dependency allows to overwrite some mods if they are enabled.

Comment (kdmcser) [reply config/schemas/mod.json]: I add more texts to explain it.

File: lib/modding/CModHandler.cpp [hunk 5]
```diff
@@ -130,9 +137,11 @@ std::vector <TModID> CModHandler::validateAndSortDependencies(std::vector <TModI
 		if(!resolvedOnCurrentTreeLevel.empty())
 		{
 			resolvedModIDs.insert(resolvedOnCurrentTreeLevel.begin(), resolvedOnCurrentTreeLevel.end());
-					continue;
+			for(auto it = resolvedOnCurrentTreeLevel.begin(); it != resolvedOnCurrentTreeLevel.end(); it++)
+				notResolvedModIDs.erase(*it);
```
Comment (IvanSavenko) [inline lib/modding/CModHandler.cpp:141]:
```suggestion
			for(const auto & it : resolvedOnCurrentTreeLevel)
				notResolvedModIDs.erase(it);
```

File: lib/modding/CModHandler.cpp [hunk 6]
```diff
@@ -158,21 +167,42 @@ std::vector <TModID> CModHandler::validateAndSortDependencies(std::vector <TModI
...
 		{
 			if (vstd::contains(allMods.at(reverseConflict).conflicts, brokenModID))
+			{
 				addErrorMessage("vcmi.server.errors.modConflict", brokenModID, reverseConflict);
+				showErrorMessage = true;
+			}
 		}
+
+		// some mods may in a (soft) dependency loop.
+		if(!showErrorMessage && brokenMod.config["modType"].String() != "Compatibility")
```
Comment (IvanSavenko) [inline lib/modding/CModHandler.cpp:197]: Why we need to exclude compatibility mods here?

File: lib/modding/ContentTypeHandler.cpp [hunk 7]
```diff
@@ -201,8 +201,10 @@ void ContentTypeHandler::afterLoadFinalization()
 			for (auto const & conflictModEntry: conflictModData.Struct())
 				conflictingMods.insert(conflictModEntry.first);
 
-			for (auto const & modID : conflictingMods)
+			for (auto const & modID : conflictingMods) {
```
Comment (IvanSavenko) [inline lib/modding/ContentTypeHandler.cpp:204]:
```suggestion
			for (auto const & modID : conflictingMods)
			{
```

File: lib/modding/CModHandler.cpp (hunk 5 above)
Comment (kdmcser) [reply lib/modding/CModHandler.cpp]: it is the Iterator of resolvedOnCurrentTreeLevel，is it safe in notResolvedModIDs.erase(it);？I dont know. So I follow your suggestion.

**Pull request merged**